datacube/utils/geometry/tools.py

```diff
@@ -446,20 +446,24 @@ def box_overlap(src_shape, dst_shape, ST, tol):
                       direction is: Xsrc = ST*Xdst
     :param       tol: Sub-pixel translation tolerance that's scaled by resolution.
     """
-    from datacube.utils.math import is_almost_int
+    from datacube.utils.math import maybe_int
 
     (sx, _, tx,
      _, sy, ty,
      *_) = ST
 
-    def _round_to_near_int(x, tol):
-        tol = min(0.5, tol)
-        if is_almost_int(x, tol):
-            x = round(x)
-        return x
+    def snap_scale(s, tol=1e-6):
+        if s > 1:
+            s = maybe_int(s, tol)
+        elif s < 1:
+            s = 1.0/maybe_int(1.0/s, tol)
```

Comment:
This needs to be done differently, it should return exactly the same value as passed in when not snapping to a fraction.

`maybe_int` returns exactly the same object when snapping is not happening, so if `maybe_int(inv_s) is inv_s` then you just return `s`, and only if `inv_s` was modified do you change `s`.